safe/_pattern.py

```diff
@@ -0,0 +1,494 @@
+from typing import Union
+from typing import List
+from typing import Optional
+import re
+import datamol as dm
+import torch
+import numpy as np
+import safe as sf
+import torch.nn.functional as F
+import transformers
+
+from loguru import logger
+from tqdm.auto import tqdm
+import contextlib
+
+
```

Comment:
Could you provide a few bullet points on the generation logic for both classes to make the explanation easier to understand?